src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
+        )
+        for operand in operands
+    ]
+
+    # keep track of the `partition` and `system_input` of the network
+    _, contracrtion_list = np.einsum_path(
+        subscripts, *tensors, optimize=False, einsum_call=True
+    )
+
+    inds, idx_rm, einsum_str, remaining, blas = contracrtion_list[0]
+    input_str, results_index = einsum_str.split("->")
+    inputs = input_str.split(",")
+
+    # Warning if the same index connects two input or two output systems
+    if not surpress_warning:
+        for ind in idx_rm:
+            found = 0
+            for i, script in enumerate(inputs):
+                try:
+                    index = script.index(ind)
+                    found += 1
+                    if found > 1:
+                        if is_input != operands[inds[i]].system_input[index]:
+                            pass
+                        else:
+                            warning(
+                                f"Index {ind} connects two {'input' if is_input else 'output'} systems."
+                            )
+                    is_input = operands[inds[i]].system_input[index]
+                    if found > 2:
+                        warning(
+                            f"Index {ind} appears multiple times in the input subscripts {input_str}."
+                        )
+                except:
+                    continue
+
+    # set correct order of the `partition` and `system_input`
+    partition = []
+    system_input = []
+    for ind in results_index:
+        for i, script in enumerate(inputs):
+            try:
+                index = script.index(ind)
+
+                partition.append(operands[inds[i]].partition[index])
+                system_input.append(operands[inds[i]].system_input[index])
+
+            except:
+                continue
+
+    new_tensor = np.einsum(subscripts, *tensors)
+
+    return QuantumNetwork(new_tensor, partition, system_input, backend=backend)
+
+
+def identity(dim: int, backend=None):
+    """Returns the identity channel.
+
+    Args:
+        dim (int): Dimension of the identity operator.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Backend to be used
+            to return the identity operator. If ``None``, defaults to the backend defined
+            when initializing the :class:`qibo.quant
+    """
+
+    return QuantumChannel.from_nparray(
+        np.eye(dim), [dim, dim], pure=True, backend=backend
+    )
+
+
+def trace(dim: int, backend=None):
+    """Returns the trace operator.
+
+    Args:
+        dim (int): Dimension to be traced.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Backend to be used
+            to return the identity operator. If ``None``, defaults to the backend defined
+            when initializing the :class:`qibo.quant
+    """
+
+    return QuantumNetwork.from_nparray(
+        np.eye(dim), [dim], [True], pure=False, backend=backend
+    )
```

Comment:
This seems like it should be a method of the `QuantumNetwork` class. Also, in principle, this should allow for multiple dimensions, no?

Reply:
Both the `identity` and the `trace` are special cases of `QuantumNetwork`. They are important tools in constructing quantum networks. So I provide them here for convenience.

Reply:
It seems like the best way to implement them would be an `IdentityNetwork` class and a `TraceNetwork` class, both inheriting from `QuantumNetwork`